### ISSUE TITLE
Can't validate presence of a serialized AR attribute

### ISSUE THREAD

**Issue by jc00ke** [thoughtbot/shoulda-matchers#757]
```ruby
# app/models/widget.rb

class Widget < ActiveRecord::Base
  validates :parts, presence: true
  serialize :parts, Array
end

# spec/models/widget_spec.rb
RSpec.describe Widget do
  it { is_expected.to validate_presence_of(:parts) }
  it { is_expected.to serialize(:parts).as(Array) }
end
```

```
  1) Widget should require parts to be set
     Failure/Error: it { is_expected.to validate_presence_of :parts }
     Shoulda::Matchers::ActiveModel::AllowValueMatcher::CouldNotSetAttributeError:
       Expected Class to be able to set parts to nil, but got [] instead.
     # /home/jesse/.gem/ruby/2.1.5/gems/shoulda-matchers-3.0.0.rc1/lib/shoulda/matchers/active_model/allow_value_matcher.rb:332:in `ensure_that_attribute_has_been_changed_to_or_from_nil!'
     # /home/jesse/.gem/ruby/2.1.5/gems/shoulda-matchers-3.0.0.rc1/lib/shoulda/matchers/active_model/allow_value_matcher.rb:315:in `set_attribute_ignoring_range_errors'
     # /home/jesse/.gem/ruby/2.1.5/gems/shoulda-matchers-3.0.0.rc1/lib/shoulda/matchers/active_model/allow_value_matcher.rb:309:in `set_attribute'
     # /home/jesse/.gem/ruby/2.1.5/gems/shoulda-matchers-3.0.0.rc1/lib/shoulda/matchers/active_model/allow_value_matcher.rb:304:in `value_matches?'
     # /home/jesse/.gem/ruby/2.1.5/gems/shoulda-matchers-3.0.0.rc1/lib/shoulda/matchers/active_model/allow_value_matcher.rb:264:in `block in matches?'
     # /home/jesse/.gem/ruby/2.1.5/gems/shoulda-matchers-3.0.0.rc1/lib/shoulda/matchers/active_model/allow_value_matcher.rb:264:in `each'
     # /home/jesse/.gem/ruby/2.1.5/gems/shoulda-matchers-3.0.0.rc1/lib/shoulda/matchers/active_model/allow_value_matcher.rb:264:in `all?'
     # /home/jesse/.gem/ruby/2.1.5/gems/shoulda-matchers-3.0.0.rc1/lib/shoulda/matchers/active_model/allow_value_matcher.rb:264:in `matches?'
     # /home/jesse/.gem/ruby/2.1.5/gems/shoulda-matchers-3.0.0.rc1/lib/shoulda/matchers/active_model/disallow_value_matcher.rb:16:in `matches?'
     # /home/jesse/.gem/ruby/2.1.5/gems/shoulda-matchers-3.0.0.rc1/lib/shoulda/matchers/active_model/validation_matcher.rb:56:in `disallows_value_of'
     # /home/jesse/.gem/ruby/2.1.5/gems/shoulda-matchers-3.0.0.rc1/lib/shoulda/matchers/active_model/validate_presence_of_matcher.rb:126:in `matches?'
     # ./spec/models/widget_spec.rb:1:in `block (2 levels) in <top (required)>'
```

I don't see anything immediately obvious in how I might tell `validate_presence_of` to use `[]` instead of `nil`.

**Comment by jc00ke** [comment]
OK, so I found the answer before I created this issue but figured others might find it useful.

See [this comment](https://github.com/thoughtbot/shoulda-matchers/blob/a3fb06c369c2b5e96831cb2c6c8532de8e2f0e62/lib/shoulda/matchers/active_model/validate_presence_of_matcher.rb#L52-L56) for an explanation.

```
# This happens because `has_secure_password` itself overrides your model
# so that it is impossible to set `password` to nil. This means that it is
# impossible to test that setting `password` to nil places your model in
# an invalid state (which in turn means that the validation itself is
# unnecessary)
```

I guess `serialize` could be treated the same way. If this is incorrect, I'll be happy to re-open.

**Comment by jc00ke** [comment]
After thinking about it, this should work. I'll try to find some time for a PR.

**Comment by mcmire** [comment]
Yeah, the issue here is that the `validate_presence_of` matcher sets the attribute in question to nil and assumes it stays nil. However, in the case of a serialized attribute, nil is actually being typecast to `[]`. This isn't actually bad in this case, because all the attribute has to be is some kind of blank. One possible solution that comes to mind is to rescue CouldNotSetAttributeError and inspect what the real value of the attribute is -- if it's `.blank?`, we're good; if not, then re-raise the error.

**Comment by jc00ke** [comment]
Before I start working on this, can I get some guidance on how to run the specs? I don't see any instructions on how I might run just the `serialize_matcher_spec.rb`. Thanks!

**Comment by maurogeorge** [comment]
@jc00ke Hi,

to run the specs we use [Appraisal](https://github.com/thoughtbot/appraisal). So to run a single spec run `$ appraisal 4.2 rspec spec/unit/shoulda/matchers/active_record/serialize_matcher_spec.rb`

More info on  [CONTRIBUTING file](https://github.com/thoughtbot/shoulda-matchers/blob/master/CONTRIBUTING.md). Hope that's help you.

**Comment by mcmire** [comment]
This makes me think that I should probably add a way to get around a CouldNotSetAttributeError if it's ever raised, perhaps as another qualifier to `allow_value`, such as `and_allow_value_to_be_changed_to` or something like that.

**Comment by mcmire** [comment]
@jc00ke This should now be fixed in 4b941254fb32840eaabaec68d41e939fe26fce04. Would you mind switching your project to use shoulda-matchers master and let me know if that fixes your test?

**Comment by jc00ke** [comment]
Nice work @mcmire, it works!
